### ISSUE TITLE
Better movement of sub-menu items in treeview.

### ISSUE THREAD

**Issue by choobin** [choobin/cmis#22]
Kevin mentioned in a previous issue.

> Another thing I just noticed when trying to re-arrange the menu after importing. I had to click UP/DOWN to move the root folder back into a root position, clicking multiple times. Could you add a button on the right side "Make Root", <-- or something? So it jumps to the main root position? Or something similar to this at least.
> 
> OR much better if possible, the ability for the user to Drag the folder into the position they want?

Indeed. There really needs to be a few more options for movement of large sub-menus. I was thinking of copy/paste/make root right-click context menu items (or buttons).

Thinking out loud. There is a slight issue with 'make root' option. What if we want to make nested sub-menu a root sub-menu. Do we place it above or bellow the current block?

**Comment by nhantrn** [comment]
You should make the tree collapsible and add drag-n-drop to manipulate the items order easier. That's the standard UI practice in Windows anyway, would be a good practice to adopt that.

**Comment by choobin** [comment]
> You should make the tree collapsible and add drag-n-drop to manipulate the items order easier.

I completely agree. Collapsing sub-menu's should relatively easy to implement. However, it might take some time to get drag and drop working. Working with a custom XUL treeview element is a real pain to work with.

**Comment by choobin** [comment]
I had an awful lot of fun over the weekend playing with the Treeview element. It should support submenu expand, collapse and dran-n-drop operations. 

Refer to the initial commit [8a23124f78](https://github.com/choobin/cmis/commit/8a23124f782e70c831c0484723f3742a4ebb64da). 

For testing I have made a development tag [20130204-dev](https://github.com/choobin/cmis/archive/20130204-dev.zip).

**Comment by KevinSanders** [comment]
LOL "Awful" and "Fun" in the same sentence doesn't sound that fun. Thanks for your time on this. Really appreciated!

I like the expand/collapse all. Nice addition.

I tried dragging some items around. But when I exported the settings, deleted all the current settings (paths list) and re-imported the saved settings file. The list was the same as before I moved things around. Otherwise the tree items are collapsing fine and moving things about is perfect.

UPDATE:
- Just noticed when things are moved about in the tree and then exit the settings dialogue, their new positions aren't saved.
- When I move an item about and drop it between some other items in a sub-menu, the item below the dropped item is selected. Whereas I was expecting the dropped item to remain selected.

**Comment by choobin** [comment]
> Just noticed when things are moved about in the tree and then exit the settings dialogue, their new positions aren't saved.

Drats. I forgot to call update after a drop operation. This -- [190b82f737](https://github.com/choobin/cmis/commit/190b82f737dbebaf64ef6ea695260de40a8228db) -- should fix it. I remade the [20130204-dev](https://github.com/choobin/cmis/archive/20130204-dev.zip) tag. It should import/export/update correctly now. I will re-open this issue until I get some time to test it further.

> When I move an item about and drop it between some other items in a sub-menu, the item below the dropped item is selected. Whereas I was expecting the dropped item to remain selected.

I encountered this too. It has to do with how drag and drop works on XUL elements. There is not much I can do in this case.

Basically you either drop before or after a selected index. Firefox seems to select a strange index and change the orientation field sometimes. With a little practice I managed to get it working as I expect. It is all about how you enter an item and where you are hovering, however, this is not ideal.

**Comment by KevinSanders** [comment]
That's done the trick! Nice one! :+1:

**Comment by choobin** [comment]
OK. So I have -- [3866312531](https://github.com/choobin/cmis/commit/38663125310f411921c739f3a4eb9dcdd48f9cd5) -- simplified the drag and drop behavior to only allow DROP_AFTER orientation events. This makes things significantly easier to deal with code wise and also fixed the select issue discussed above. I am still going to keep this open for a while to see how it goes.

**Comment by choobin** [comment]
Drag and drop, expand/collapse and correct selection of items after movement all work without a hitch. I am quite happy with it.

Refer to [20130207](https://github.com/choobin/cmis/archive/20130207.zip). I have actually submitted this version to AMO. It should be live within 1~2 weeks.

**Comment by nhantrn** [comment]
great job! thanks for your hard work.
